File: strings/08_valid_word_abbreviation/test_valid_word_abbreviation.py
from unittest import TestCase, TestResult, TestLoader, TextTestRunner
from valid_word_abbreviation import Solution
from time import time

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


@time_tests
class TestValidWordAbbreviation(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test1_i12iz4n(self):
        word = "internationalization"
        abbr = "i12iz4n"
        expected = True
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test2_a2e(self):
        word = "apple"
        abbr = "a2e"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test3_a(self):
        word = "a"
        abbr = "2"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test4_s10n(self):
        word = "substitution"
        abbr = "s10n"
        expected = True
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test5_sub4u4(self):
        word = "substitution"
        abbr = "sub4u4"
        expected = True
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test6_12(self):
        word = "substitution"
        abbr = "12"
        expected = True
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test8_substitution(self):
        word = "substitution"
        abbr = "substitution"
        expected = True
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test9_s55n(self):
        word = "substitution"
        abbr = "s55n"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test10_s010n(self):
        word = "substitution"
        abbr = "s010n"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test11_s0ubstitution(self):
        word = "substitution"
        abbr = "s0ubstitution"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test12_hi1(self):
        word = "hi"
        abbr = "hi1"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def test13_1(self):
        word = "hi"
        abbr = "1"
        expected = False
        result = self.solution.valid_word_abbreviation(word, abbr)

        self.assertEqual(result, expected)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestValidWordAbbreviation)
    result = CustomTestResult()
    runner = TextTestRunner(verbosity=2)
    # runner.run(suite)
    suite.run(result)
    result.printTestTimes()

File: strings/08_valid_word_abbreviation/test_valid_word_abbreviation.cpp
#include "valid_word_abbreviation.cpp"
#include <gtest/gtest.h>
#include <string>
#include <chrono>
#include <iostream>
 
using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestValidWordAbbreviation : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(TestValidWordAbbreviation, test1_i12iz4n) {
    string word = "internationalization";
    string abbr = "i12iz4n";
    bool expected = true;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test2_a2e) { 
    string word = "apple";
    string abbr = "a2e";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test3_a) { 
    string word = "a";
    string abbr = "2";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test4_s10n) { 
    string word = "substitution";
    string abbr = "s10n";
    bool expected = true;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test5_sub4u4) { 
    string word = "substitution";
    string abbr = "sub4u4";
    bool expected = true;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test6_12) { 
    string word = "substitution";
    string abbr = "12";
    bool expected = true;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test7_su3i1u2on) { 
    string word = "substitution";
    string abbr = "su3i1u2on";
    bool expected = true;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test8_substitution) { 
    string word = "substitution";
    string abbr = "substitution";
    bool expected = true;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test9_s55n) { 
    string word = "substitution";
    string abbr = "s55n";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test10_s010n) { 
    string word = "substitution";
    string abbr = "s010n";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test11_s0ubstitution) { 
    string word = "substitution";
    string abbr = "s0ubstitution";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test12_hi1) { 
    string word = "hi";
    string abbr = "hi1";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

TEST_F(TestValidWordAbbreviation, test13_1) { 
    string word = "hi";
    string abbr = "1";
    bool expected = false;
    bool result = solution->validWordAbbreviation(word, abbr);

    ASSERT_EQ(expected, result);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}